### PR TITLE
Pack instructions and state more efficiently

### PR DESCRIPTION
**Pull request by joncinque**
* SwapInfo and SwapInstruction are more memory-efficient, following the
  example from SPL Token
* Remove unnecessary helper structs to make instructions clearer
* Error -> SwapError
* Update JS tests to reflect the change

### CODE REVIEW

File: token-swap/program/src/instruction.rs [hunk 1]
```diff
@@ -305,29 +311,53 @@ mod tests {
...
             fee_numerator,
             fee_denominator,
             nonce,
-        });
-        let packed = check.serialize().unwrap();
-        let unpacked = SwapInstruction::deserialize(&packed).unwrap();
+        };
+        let packed = check.pack();
+        println!("{:?}", packed);
+        let unpacked = SwapInstruction::unpack(&packed).unwrap();
```
Comment (CriesofCarrots) [inline token-swap/program/src/instruction.rs:325]: It doesn't look like we compare this `packed` against the expected `data` below. I like the patterns for the other instructions better.
Also axe that `println` 😄

File: token-swap/program/src/processor.rs [hunk 2]
```diff
@@ -34,19 +34,23 @@ const TOKEN_PROGRAM_ID: Pubkey = Pubkey::new_from_array([1u8; 32]);
 pub struct Processor {}
 impl Processor {
     /// Deserializes a spl_token `Account`.
-    pub fn token_account_deserialize(data: &[u8]) -> Result<spl_token::state::Account, Error> {
-        spl_token::state::Account::unpack(data).map_err(|_| Error::ExpectedAccount)
+    pub fn token_account_deserialize(data: &[u8]) -> Result<spl_token::state::Account, SwapError> {
```
Comment (CriesofCarrots) [inline token-swap/program/src/processor.rs:37]: Now that we're using `unpack` everywhere, maybe rename these methods?
Suggest: `unpack_token_account` and `unpack_mint`

File: token-swap/program/src/processor.rs [hunk 3]
```diff
@@ -218,23 +227,26 @@ impl Processor {
         let destination_info = next_account_info(account_info_iter)?;
         let token_program_info = next_account_info(account_info_iter)?;
 
-        let token_swap = State::deserialize(&swap_info.data.borrow())?.token_swap()?;
+        let token_swap = SwapInfo::unpack(&swap_info.data.borrow())?;
+        if !token_swap.is_initialized {
+            return Err(SwapError::AlreadyInUse.into());
+        }
```
Comment (CriesofCarrots) [inline token-swap/program/src/processor.rs:233]: We do this check several times; maybe time for an `unpack_checked` method? (Or an `unpack_unchecked` method for the init instruction, if you prefer)

Comment (CriesofCarrots) [reply token-swap/program/src/processor.rs:233]: Also, should this be `SwapError::InvalidState`?

File: token-swap/js/client/token-swap.js [hunk 4]
```diff
@@ -97,15 +97,13 @@ type TokenSwapInfo = {|
  * @private
  */
 const TokenSwapLayout = BufferLayout.struct([
-  BufferLayout.u8('state'),
+  BufferLayout.u8('is_initialized'),
```
Comment (CriesofCarrots) [inline token-swap/js/client/token-swap.js:100]: Camel-case in js, please

File: token-swap/js/client/token-swap.js [hunk 5]
```diff
@@ -273,7 +270,7 @@ export class TokenSwap {
 
     const data = Buffer.from(accountInfo.data);
     const tokenSwapInfo = TokenSwapLayout.decode(data);
-    if (tokenSwapInfo.state !== 1) {
+    if (!tokenSwapInfo.is_initialized) {
```
Comment (CriesofCarrots) [inline token-swap/js/client/token-swap.js:273]: Ditto (camel-case)

**Review by CriesofCarrots — COMMENTED** [review]
Looking good! Just a few small suggestions

File: token-swap/js/client/token-swap.js (hunk 4 above)
Comment (joncinque) [reply token-swap/js/client/token-swap.js:100]: Oops, thanks for the catch

File: token-swap/program/src/instruction.rs (hunk 1 above)
Comment (joncinque) [reply token-swap/program/src/instruction.rs:325]: That's embarrassing :sweat_smile: and that's right, the check was never done.  I'll make it look like the other tests

File: token-swap/program/src/processor.rs (hunk 2 above)
Comment (joncinque) [reply token-swap/program/src/processor.rs:37]: Definitely

File: token-swap/program/src/processor.rs (hunk 3 above)
Comment (joncinque) [reply token-swap/program/src/processor.rs:233]: I'll follow the `unpack_unchecked` model from `Token`, and absolutely, `InvalidState` is a better error

**Review by CriesofCarrots — APPROVED** [review]
Nice!

**Pull request merged**